File: backend/Testing/Browser/Page/UserDashboardPage.cs
using Microsoft.Playwright;
using Shouldly;

namespace Testing.Browser.Page;

public class UserDashboardPage : AuthenticatedBasePage<UserDashboardPage>
{
    public UserDashboardPage(IPage page)
    : base(page, "/", page.GetByRole(AriaRole.Heading, new() { Name = "My Projects" }))
    {
    }

    public async Task<ProjectPage> OpenProject(string projectName, string projectCode)
    {
        var projectHeader = Page.GetByRole(AriaRole.Heading, new() { Name = projectName });
        var projectCard = Page.Locator(".card", new() { Has = projectHeader });
        projectCode.ShouldNotBeNullOrEmpty();
        await projectCard.ClickAsync();
        return await new ProjectPage(Page, projectName, projectCode).WaitFor();
    }
}

File: frontend/tests/pages/userDashboardPage.ts
import {AuthenticatedBasePage} from './authenticatedBasePage';
import {CreateProjectPage} from './createProjectPage';
import type {Page} from '@playwright/test';
import {ProjectPage} from './projectPage';

export class UserDashboardPage extends AuthenticatedBasePage {
  constructor(page: Page) {
    super(page, page.getByRole('heading', {name: 'My Projects'}), '/');
  }

  async openProject(projectName: string, projectCode: string): Promise<ProjectPage> {
    await this.selectGridView();
    const projectHeader = this.page.getByRole('heading', {name: projectName});
    const projectCard = this.page.locator('.card', {has: projectHeader});
    await projectCard.click();
    return new ProjectPage(this.page, projectName, projectCode).waitFor();
  }

  async selectGridView(): Promise<void> {
    await this.page.click('.i-mdi-grid');
  }

  async clickCreateProject(): Promise<CreateProjectPage> {
    await this.page.getByRole('link', {name: /(Create|Request) Project/, exact: true}).click();
    return new CreateProjectPage(this.page).waitFor();
  }
}
